sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -0,0 +1,245 @@
+):
+
+    cdef:
+        cnp.ndarray[cnp.double_t, ndim=1] current_distances_arr
+        cnp.ndarray[cnp.double_t, ndim=1] current_sources_arr
+        cnp.ndarray[cnp.int8_t, ndim=1] in_tree_arr
+        cnp.ndarray[cnp.double_t, ndim=2] result_arr
+
+        cnp.double_t * current_distances
+        cnp.double_t * current_sources
+        cnp.double_t * current_core_distances
+        cnp.double_t * raw_data_ptr
+        cnp.int8_t * in_tree
+        cnp.double_t[:, ::1] raw_data_view
+        cnp.double_t[:, ::1] result
+
+        cnp.ndarray label_filter
+
+        cnp.intp_t current_node
+        cnp.intp_t source_node
+        cnp.intp_t right_node
+        cnp.intp_t left_node
+        cnp.intp_t new_node
+        cnp.intp_t i
+        cnp.intp_t j
+        cnp.intp_t dim
+        cnp.intp_t num_features
+
+        double current_node_core_distance
+        double right_value
+        double left_value
+        double core_value
+        double new_distance
+
+    dim = raw_data.shape[0]
+    num_features = raw_data.shape[1]
+
+    raw_data_view = (<cnp.double_t[:raw_data.shape[0], :raw_data.shape[1]:1]> (
+        <cnp.double_t *> raw_data.data))
+    raw_data_ptr = (<cnp.double_t *> &raw_data_view[0, 0])
+
+    result_arr = np.zeros((dim - 1, 3))
+    in_tree_arr = np.zeros(dim, dtype=np.int8)
+    current_node = 0
+    current_distances_arr = np.infty * np.ones(dim)
+    current_sources_arr = np.ones(dim)
+
+    result = (<cnp.double_t[:dim - 1, :3:1]> (<cnp.double_t *> result_arr.data))
+    in_tree = (<cnp.int8_t *> in_tree_arr.data)
+    current_distances = (<cnp.double_t *> current_distances_arr.data)
+    current_sources = (<cnp.double_t *> current_sources_arr.data)
+    current_core_distances = (<cnp.double_t *> core_distances.data)
+
+    for i in range(1, dim):
+
+        in_tree[current_node] = 1
+
+        current_node_core_distance = current_core_distances[current_node]
+
+        new_distance = DBL_MAX
+        source_node = 0
+        new_node = 0
+
+        for j in range(dim):
+            if in_tree[j]:
+                continue
+
+            right_value = current_distances[j]
+            right_source = current_sources[j]
+
+            left_value = dist_metric.dist(&raw_data_ptr[num_features *
+                                                        current_node],
+                                          &raw_data_ptr[num_features * j],
+                                          num_features)
+            left_source = current_node
+
+            if alpha != 1.0:
+                left_value /= alpha
+
+            core_value = core_distances[j]
+            if (current_node_core_distance > right_value or
+                    core_value > right_value or
+                    left_value > right_value):
+                if right_value < new_distance:
+                    new_distance = right_value
+                    source_node = right_source
+                    new_node = j
+                continue
```

Comment:
If this is the same as the algorithm above, we have a pairwise-computation + argmin. It looks like the chunking and work of @jjerphan should be useful here.